#### 1. Connecting with database
First lines enable connection to database created in MySQL:

In [1]:
from sqlalchemy import create_engine

In [2]:
engine = create_engine('postgresql://root:root@localhost:5432/pl_data')

In [12]:
%load_ext sql

%sql postgresql://root:root@localhost:5432/pl_data

The sql extension is already loaded. To reload it, use:
  %reload_ext sql


#### 2. Looking at the available data in database

In [14]:
%sql select * from premier_league_stats limit 5

 * postgresql://root:***@localhost:5432/pl_data
5 rows affected.


[(0, 1993, 1, '1992-08-15', 'Coventry City', 2, 1, 'Middlesbrough', 'H'),
 (1, 1993, 1, '1992-08-15', 'Leeds United', 2, 1, 'Wimbledon', 'H'),
 (2, 1993, 1, '1992-08-15', 'Sheffield Utd', 2, 1, 'Manchester Utd', 'H'),
 (3, 1993, 1, '1992-08-15', 'Crystal Palace', 3, 3, 'Blackburn', 'D'),
 (4, 1993, 1, '1992-08-15', 'Arsenal', 2, 4, 'Norwich City', 'A')]

#### 3. Checking if the season data is complete - each team should have played twice against their opponents (a home match and an away match)

In [17]:
%%sql 
select 
	"Season_End_Year", 
    count("Season_End_Year") as Matches_Played
from premier_league_stats
group by "Season_End_Year"
Order by "Season_End_Year";

 * postgresql://root:***@localhost:5432/pl_data
31 rows affected.


[(1993, 462),
 (1994, 462),
 (1995, 462),
 (1996, 380),
 (1997, 380),
 (1998, 380),
 (1999, 380),
 (2000, 380),
 (2001, 380),
 (2002, 380),
 (2003, 380),
 (2004, 380),
 (2005, 380),
 (2006, 380),
 (2007, 380),
 (2008, 380),
 (2009, 380),
 (2010, 380),
 (2011, 380),
 (2012, 380),
 (2013, 380),
 (2014, 380),
 (2015, 380),
 (2016, 380),
 (2017, 380),
 (2018, 380),
 (2019, 380),
 (2020, 380),
 (2021, 380),
 (2022, 380),
 (2023, 380)]

The Premier League was reduced to *20 teams* starting from the *1995/1996* season. 
However, the first three seasons of the Premier League were played with *22 teams* in the league.

#### 4. Selecting seasons stats - results: home wins, draws, away wins and as well as percentage for each result type

Below an example for highest wins percentage - using **ORDER BY** statement:

In [18]:
%%sql
select 
	"Season_End_Year", 
    count("Season_End_Year") as matches_count,

    count(CASE WHEN "FTR" = 'H' THEN "FTR" END) as Home_Wins, 
    (cast(count(CASE WHEN "FTR" = 'H' THEN "FTR" END) as decimal(10,2))/cast(count("Season_End_Year") as decimal(10,2)))*100 as Home_Wins_Percentage,
    
    count(CASE WHEN "FTR" = 'D' THEN "FTR" END) as Draws, 
    (cast(count(CASE WHEN "FTR" = 'D' THEN "FTR" END) as decimal(10,2))/cast(count("Season_End_Year") as decimal(10,2)))*100 as Draws_Percentage,
    
    count(CASE WHEN "FTR" = 'A' THEN "FTR" END) as Away_Wins, 
    (cast(count(CASE WHEN "FTR" = 'A' THEN "FTR" END) as decimal(10,2))/cast(count("Season_End_Year") as decimal(10,2)))*100 as Away_Wins_Percentage

from premier_league_stats
group by "Season_End_Year"
order by  Home_Wins_Percentage DESC
limit 5;

 * postgresql://root:***@localhost:5432/pl_data
5 rows affected.


[(2010, 380, 193, Decimal('50.78947368421052631600'), 96, Decimal('25.26315789473684210500'), 91, Decimal('23.94736842105263157900')),
 (2006, 380, 192, Decimal('50.52631578947368421100'), 77, Decimal('20.26315789473684210500'), 111, Decimal('29.21052631578947368400')),
 (2000, 380, 187, Decimal('49.21052631578947368400'), 92, Decimal('24.21052631578947368400'), 101, Decimal('26.57894736842105263200')),
 (2017, 380, 187, Decimal('49.21052631578947368400'), 84, Decimal('22.10526315789473684200'), 109, Decimal('28.68421052631578947400')),
 (2003, 380, 187, Decimal('49.21052631578947368400'), 90, Decimal('23.68421052631578947400'), 103, Decimal('27.10526315789473684200'))]

The highest percentage of *home wins* occurred during the *2009/2010* season, followed by the *2005/2006* season, with both seasons having a percentage *above 50%*.
Another interesting thing to note - during the *2018/2019* season, there were *less than 20%* of *draws*, specifically 18.68%.
Additionally, the season with the highest percentage of *away wins* was the *2020/2021* season, with *over 40%* of matches being won by the away team, which is significantly higher than second-highest away wins percentage of *33.95%** in *2021/2022*. 

#### 5. Selecting seasons stats - goals scored and goals per game for each season

Below an example for highest goals per game rate - using **ORDER BY** statement:

In [19]:
%%sql
select 
	"Season_End_Year", sum("HomeGoals" + "AwayGoals") as Goals_Scored, 
    cast(sum("HomeGoals" + "AwayGoals")/count("Season_End_Year") as decimal (10,2)) as Goals_Per_Game
from premier_league_stats
group by "Season_End_Year"
order by Goals_Per_Game DESC
limit 5;

 * postgresql://root:***@localhost:5432/pl_data
5 rows affected.


[(2023, Decimal('1084'), Decimal('2.85')),
 (2019, Decimal('1072'), Decimal('2.82')),
 (2022, Decimal('1071'), Decimal('2.82')),
 (2012, Decimal('1066'), Decimal('2.81')),
 (2011, Decimal('1063'), Decimal('2.80'))]

During the *2018/2019* and *2021/2022* seasons, the highest average number of goals scored per game was *2.82*. 
On the other hand, the lowest average number of goals scored per game occurred during the *2006/2007* season, with an average of *2.45* goals per game.

#### 6. Selecting goals stats
Selecting all goals (Home & Away) by each team for all the seasons to get the most scoring teams since the premier league started.

In [21]:
%%sql
SELECT Team, SUM(Goals) AS TotalGoals
FROM (
    SELECT "Home" AS Team, "HomeGoals" AS Goals
    FROM premier_league_stats
    UNION ALL
    SELECT "Away" AS Team, "AwayGoals" AS Goals
    FROM premier_league_stats
) AS Combined
GROUP BY Team
ORDER BY TotalGoals DESC
limit 5;

 * postgresql://root:***@localhost:5432/pl_data
5 rows affected.


[('Manchester Utd', Decimal('2243')),
 ('Arsenal', Decimal('2105')),
 ('Liverpool', Decimal('2096')),
 ('Chelsea', Decimal('2011')),
 ('Tottenham', Decimal('1815'))]

#### 7. Creating new table with home match stats for each team

To get the same order for both home wins and away teams tables (next step) all results were sorted by team name and season.

In [25]:
%%sql
CREATE table home_wins As
select 
	"Season_End_Year",
    "Home",
    count(CASE WHEN "FTR" = 'H' THEN "FTR" END) as HomeWins,
    count(CASE WHEN "FTR" = 'D' THEN "FTR" END) as HomeDraws,
    count(CASE WHEN "FTR" = 'A' THEN "FTR" END) as HomeLoses,
    sum("HomeGoals") as HomeGoalsScored,
    sum("AwayGoals") as HomeGoalsConceded
from premier_league_stats
group by "Season_End_Year", "Home"
order by "Home", "Season_End_Year";

 * postgresql://root:***@localhost:5432/pl_data
626 rows affected.


[]

Checking most home wins over all seasons:

In [26]:
%%sql
select * from home_wins
order by HomeWins DESC
limit 6;

 * postgresql://root:***@localhost:5432/pl_data
6 rows affected.


[(2012, 'Manchester City', 18, 1, 0, Decimal('55'), Decimal('12')),
 (2011, 'Manchester Utd', 18, 1, 0, Decimal('49'), Decimal('12')),
 (2020, 'Liverpool', 18, 1, 0, Decimal('52'), Decimal('16')),
 (2006, 'Chelsea', 18, 1, 0, Decimal('47'), Decimal('9')),
 (2019, 'Manchester City', 18, 0, 1, Decimal('57'), Decimal('12')),
 (2017, 'Chelsea', 17, 0, 2, Decimal('55'), Decimal('17'))]

Only a few teams in Premier League history have achieved *18 out of 19 home wins* in a single season. These teams include *Man City* (twice), *Man United, Liverpool* and *Chelsea*.

#### 8. Creating new table with away match stats for each team

In [27]:
%%sql
create table away_wins As
select 
	"Season_End_Year",
    "Away",
    count(CASE WHEN "FTR" = 'A' THEN "FTR" END) as AwayWins,
    count(CASE WHEN "FTR" = 'D' THEN "FTR" END) as AwayDraws,
    count(CASE WHEN "FTR" = 'H' THEN "FTR" END) as AwayLoses,
    sum("AwayGoals") as AwayGoalsScored,
    sum("HomeGoals") as AwayGoalsConceded
from premier_league_stats
group by "Season_End_Year", "Away"
order by "Away", "Season_End_Year";

 * postgresql://root:***@localhost:5432/pl_data
626 rows affected.


[]

Checking most away wins over all seasons:

In [28]:
%%sql
select * from away_wins
order by AwayWins DESC
limit 5;

 * postgresql://root:***@localhost:5432/pl_data
5 rows affected.


[(2018, 'Manchester City', 16, 2, 1, Decimal('45'), Decimal('13')),
 (2005, 'Chelsea', 15, 3, 1, Decimal('37'), Decimal('9')),
 (2009, 'Chelsea', 14, 2, 3, Decimal('35'), Decimal('12')),
 (2002, 'Arsenal', 14, 5, 0, Decimal('37'), Decimal('11')),
 (2020, 'Liverpool', 14, 2, 3, Decimal('33'), Decimal('17'))]

Only one team in Premier League history has achieved *16 out of 19 away wins* in a single season. 
This team is *Manchester City* in 2017/2018 season.

#### 9. Creating detailed overview of team's performance over all seasons from both new created tables

**COALESCE** was used to get first non-null selection (create one column from two columns).

In [29]:
%%sql
select
    coalesce(a."Season_End_Year", h."Season_End_Year") as Season_End_Year,
    coalesce(a."Away", h."Home") as Team,
    h.HomeWins, a.AwayWins,
    h.HomeDraws, a.AwayDraws,
	h.HomeLoses, a.AwayLoses,
    h.HomeGoalsScored, a.AwayGoalsScored,
    h.HomeGoalsConceded, a.AwayGoalsConceded
from away_wins a
left join home_wins h
    on a."Season_End_Year" = h."Season_End_Year"
    and a."Away" = h."Home"
limit 5;

 * postgresql://root:***@localhost:5432/pl_data
5 rows affected.


[(1993, 'Arsenal', 8, 7, 6, 5, 7, 9, Decimal('25'), Decimal('15'), Decimal('20'), Decimal('18')),
 (1994, 'Arsenal', 10, 8, 8, 9, 3, 4, Decimal('25'), Decimal('28'), Decimal('15'), Decimal('13')),
 (1995, 'Arsenal', 6, 7, 9, 3, 6, 11, Decimal('27'), Decimal('25'), Decimal('21'), Decimal('28')),
 (1996, 'Arsenal', 10, 7, 7, 5, 2, 7, Decimal('30'), Decimal('19'), Decimal('16'), Decimal('16')),
 (1997, 'Arsenal', 10, 9, 5, 6, 4, 4, Decimal('36'), Decimal('26'), Decimal('18'), Decimal('14'))]

#### 10. Creating short overview of team's performance over all seasons from both new created tables

In [32]:
%%sql
create table perf_overview As
select
    coalesce(a."Season_End_Year", h."Season_End_Year") as Season_End_Year,
    coalesce(a."Away", h."Home") as Team,
    h.HomeWins +  a.AwayWins as Wins,
    h.HomeDraws +  a.AwayDraws  as Draws,
	h.HomeLoses +  a.AwayLoses as Loses,
    h.HomeGoalsScored + a.AwayGoalsScored as GoalsScored,
    h.HomeGoalsConceded + a.AwayGoalsConceded as GoalsConceded,
    3*(h.HomeWins +  a.AwayWins) + 1*(h.HomeDraws +  a.AwayDraws) as Points
from away_wins a
left join home_wins h
    on a."Season_End_Year" = h."Season_End_Year"
    and a."Away" = h."Home";

 * postgresql://root:***@localhost:5432/pl_data
626 rows affected.


[]

#### 11. Finding the best season for every team in PL history in terms of points gained

To get the complete overview of the best season for each team I used **subquery** that retrieves the maximum points for each team and the corresponding season.
Than I **joined** that subquery with original table.

In [33]:
%%sql
select p.*
from perf_overview as p
inner join (
  select Team, max(Points) as MaxPoints
  from perf_overview
  group by Team
) as max_points
on p.Team = max_points.Team and p.Points = max_points.MaxPoints
order by Points DESC;

 * postgresql://root:***@localhost:5432/pl_data
53 rows affected.


[(2018, 'Manchester City', 32, 4, 2, Decimal('106'), Decimal('27'), 100),
 (2020, 'Liverpool', 32, 3, 3, Decimal('85'), Decimal('33'), 99),
 (2005, 'Chelsea', 29, 8, 1, Decimal('72'), Decimal('15'), 95),
 (1994, 'Manchester Utd', 27, 11, 4, Decimal('80'), Decimal('38'), 92),
 (2004, 'Arsenal', 26, 12, 0, Decimal('73'), Decimal('26'), 90),
 (1995, 'Blackburn', 27, 8, 7, Decimal('80'), Decimal('39'), 89),
 (2017, 'Tottenham', 26, 8, 4, Decimal('86'), Decimal('26'), 86),
 (2016, 'Leicester City', 23, 12, 3, Decimal('68'), Decimal('36'), 81),
 (1996, 'Newcastle Utd', 24, 6, 8, Decimal('66'), Decimal('37'), 78),
 (1995, "Nott'ham Forest", 22, 11, 9, Decimal('72'), Decimal('43'), 77),
 (1993, 'Aston Villa', 21, 11, 10, Decimal('57'), Decimal('40'), 74),
 (1995, 'Leeds United', 20, 13, 9, Decimal('59'), Decimal('38'), 73),
 (1993, 'Norwich City', 21, 9, 12, Decimal('61'), Decimal('65'), 72),
 (2014, 'Everton', 21, 9, 8, Decimal('61'), Decimal('39'), 72),
 (2001, 'Ipswich Town', 20, 6, 12, Decimal('57'), Decimal('42'), 66),
 (2021, 'West Ham', 19, 8, 11, Decimal('62'), Decimal('47'), 65),
 (1994, 'Wimbledon', 18, 11, 13, Decimal('56'), Decimal('53'), 65),
 (1994, 'Sheffield Weds', 16, 16, 10, Decimal('76'), Decimal('54'), 64),
 (2016, 'Southampton', 18, 9, 11, Decimal('59'), Decimal('41'), 63),
 (1993, 'QPR', 17, 12, 13, Decimal('63'), Decimal('55'), 63),
 (2023, 'Brighton', 18, 8, 12, Decimal('72'), Decimal('53'), 62),
 (2020, 'Wolves', 15, 14, 9, Decimal('51'), Decimal('40'), 59),
 (2023, 'Brentford', 15, 14, 9, Decimal('58'), Decimal('46'), 59),
 (2000, 'Sunderland', 16, 10, 12, Decimal('57'), Decimal('56'), 58),
 (2005, 'Bolton', 16, 10, 12, Decimal('49'), Decimal('44'), 58),
 (2008, 'Portsmouth', 16, 9, 13, Decimal('48'), Decimal('40'), 57),
 (1994, 'Coventry City', 14, 14, 14, Decimal('43'), Decimal('45'), 56),
 (2015, 'Swansea City', 16, 8, 14, Decimal('46'), Decimal('49'), 56),
 (2005, 'Middlesbrough', 14, 13, 11, Decimal('53'), Decimal('46'), 55),
 (2007, 'Reading', 16, 7, 15, Decimal('52'), Decimal('47'), 55),
 (1998, 'Derby County', 16, 7, 15, Decimal('52'), Decimal('49'), 55),
 (2018, 'Burnley', 14, 12, 12, Decimal('36'), Decimal('39'), 54),
 (2015, 'Stoke City', 15, 9, 14, Decimal('48'), Decimal('45'), 54),
 (2020, 'Sheffield Utd', 14, 12, 12, Decimal('39'), Decimal('39'), 54),
 (2020, 'Burnley', 15, 9, 14, Decimal('43'), Decimal('50'), 54),
 (2004, 'Charlton Ath', 14, 11, 13, Decimal('51'), Decimal('51'), 53),
 (2009, 'Fulham', 14, 11, 13, Decimal('39'), Decimal('34'), 53),
 (2006, 'Wigan Athletic', 15, 6, 17, Decimal('45'), Decimal('52'), 51),
 (2004, 'Birmingham City', 12, 14, 12, Decimal('43'), Decimal('48'), 50),
 (2019, 'Watford', 14, 8, 16, Decimal('52'), Decimal('59'), 50),
 (2010, 'Birmingham City', 13, 11, 14, Decimal('38'), Decimal('47'), 50),
 (1993, 'Oldham Athletic', 13, 10, 19, Decimal('63'), Decimal('74'), 49),
 (2019, 'Crystal Palace', 14, 7, 17, Decimal('51'), Decimal('53'), 49),
 (1993, 'Crystal Palace', 11, 16, 15, Decimal('48'), Decimal('61'), 49),
 (2013, 'West Brom', 14, 7, 17, Decimal('53'), Decimal('57'), 49),
 (2017, 'Bournemouth', 12, 10, 16, Decimal('55'), Decimal('67'), 46),
 (2011, 'Blackpool', 10, 9, 19, Decimal('55'), Decimal('78'), 39),
 (2018, 'Huddersfield', 9, 10, 19, Decimal('28'), Decimal('58'), 37),
 (2014, 'Hull City', 10, 7, 21, Decimal('38'), Decimal('53'), 37),
 (2000, 'Bradford City', 9, 9, 20, Decimal('38'), Decimal('68'), 36),
 (1998, 'Barnsley', 10, 5, 23, Decimal('37'), Decimal('82'), 35),
 (2019, 'Cardiff City', 10, 4, 24, Decimal('34'), Decimal('69'), 34),
 (1994, 'Swindon Town', 5, 15, 22, Decimal('47'), Decimal('100'), 30)]

The only team in Premier League history to collect *100* points was *Manchester City* in *2017/2018* season.

Analysing this table it is possible to get some more interesting insights, e.g. less goals conceded, most goals scored etc.